### ISSUE TITLE
namespace and events handlers 

### ISSUE THREAD

**Issue by blaisezzz** [miguelgrinberg/Flask-SocketIO#428]
Hi,

What would you recommend in terms organising event handlers. Using namespace is one way of group related events together, however I am unsure how it would work out if I decide to have numbers of subclass.
For example, 
namespace
|--- VeryGenericNamespace
       |---   GenericNamespace
              |--- MyNamespace
                   |----- MyChildNamespace

Object-oriented language such as C# gives me the ability to polymorph my methods during runtime. So I can basically check my namespace type at run-time to determine what events to use.  I would also like to avoid restarting my flask application in order to switch between namespaces. It would be great if I can do everything at run-time to make it happen.

It would be awesome to know how to make my event handler extensible and modifiable using decorators and meta-programming.

Basically, I want to be able to run events dynamically depending on conditions during run-time.

**Comment by miguelgrinberg** [comment]
Namespaces are not a good way to do this. In Socket.IO a namespace represents a connection. The list of connected clients are maintained separately for each namespace, so client applications that want to access multiple namespaces will need to connect to each of them separately.

My recommendation is that you use a single namespace. I'm personally not too big on using a class to represent a namespace, but if you like that, then you can create a class-based namespace and handle your events in it. My preference would be to use a standalone function for each event handler, and if you have too many you can then group them in Python modules. If you want to have dynamic events you can use a single event name, and pass information regarding what action to carry out as arguments. In the event handler then you can implement your own dispatching.

**Comment by blaisezzz** [comment]
In this scenario, I would likely implement something like so

**Comment by blaisezzz** [comment]
In this scenario, it may look something like 

```
@socketio.on('message_event', namespace='/test')
def all_message(message):
     if message['type'] == 'private':
         # process private message 
     elif message['type'] == 'public':
         # process public message
     else: 
         # process general message 
      emit('response', {'data': process_data, 'type': message['type']})

```
That would definitely work. Have you try out other implementation?  I would like to explore alternatives if you are okay to share your insight.

**Comment by miguelgrinberg** [comment]
Yeah, that is exactly what I meant.

I have never needed to do something like this myself, I tend towards static events, so if I were to do this, I would probably define `private_message`, `public_message` and `general_message` events. But I've seen several implementations that work in this fashion.

One advice I can give if you follow this route, is to make sure that every possible message the client can send is whitelisted. For example, you definitely don't want the client to send a string with a name of a Python function that you then blindly invoke.

**Comment by blaisezzz** [comment]
That's one of the drawbacks: Need to explicitly handle every possible condition, otherwise it's prone to code breach. It's also not extensible..... I can't freely extend parts I want in future version. I am currently exploring dynamic function call and meta-programming to try to understand if that will help me or not. I will keep you updated later on. Thanks for your help.
